Makefile.build

```diff
@@ -630,15 +627,73 @@ test-suite: world byte $(ALLSTDLIB).v
 # Default rules for compiling ML code
 ###########################################################################
 
-# Target for libraries .cma and .cmxa
+gramlib__pack:
+	mkdir -p $@
+
+# gramlib.ml contents
+gramlib__pack/gramlib.ml: | gramlib__pack
+	echo " \
+module Ploc    = Gramlib__Ploc \
+module Plexing = Gramlib__Plexing \
+module Gramext = Gramlib__Gramext \
+module Grammar = Gramlib__Grammar" > $@
+
+gramlib__pack/gramlib__Ploc.mli: gramlib/ploc.mli | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Ploc.ml: gramlib/ploc.ml | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Plexing.mli: gramlib/plexing.mli | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Plexing.ml: gramlib/plexing.ml | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Gramext.mli: gramlib/gramext.mli | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Gramext.ml: gramlib/gramext.ml | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Grammar.mli: gramlib/grammar.mli | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
+gramlib__pack/gramlib__Grammar.ml: gramlib/grammar.ml | gramlib__pack
+	cp -a $< $@
+	sed -e "1i # 1 \"$<\"" -i $@
```

Comment:
I think this bunch of lines should be rewritten as follows
```
gramlib__pack/gramlib__G%: gramlib/g% | gramlib__pack
  (echo "# This file is a copy of $<"; cat $<) > $@
```
unless I'm mistaken one rule can cover all cases, and the rule body I propose needs no escaping.

Reply:
>the rule body I propose needs no escaping.

`# 1 "foo"` at the beginning of the file makes ocaml errors be printed for `foo`.

We can avoid escaping by using single quotes rather than double quotes `'# 1 "$<"'` since `$<` is expanded by make which ignores quotes. (regardless of using cp+sed or echo+cat)

Reply:
We'd need two rules I guess for `ploc -> Ploc`. Indeed we are using the mechanical form OCaml understands; I also prefer `cp` as it should correctly preserve the timestamps and attributes of the file.

Reply:
> it should correctly preserve the timestamps

Isn't `sed -i` updating the mtime?

Reply:
Let's just do `sed '1i # 1 "$<"' $< > $@`, no need for cp or cat.

Reply:
OK, but I'm starting to feel a bit dizzy.

Reply:
@SkySkimmer your idea doesn't work on OSX , so yeah, another few amends for no good reason.

Reply:
I propose two things. Factor the rules. Get rid of sed. The former is more important. For the latter my solution works, but I'm also fine with sed -i.

Reply:
What kind of rule factoring should be done?

Reply:
Isn't that already done ? I don't see what more can be improved. Cared to share the concrete part to be changed?

Reply:
@gares: well anticipated! I'm indeed meeting the issue that `sed` is different on MacOS (GNU vs BSD and POSIX style as far as I can judge). Here, option `i` needs a `\` (see [sed doc](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html)). I could not manage to make a `sed`-based rule work on MacOS, but the following workaround worked:
```
diff --git a/Makefile.build b/Makefile.build
index 8e4b63c364..0e3ced313d 100644
--- a/Makefile.build
+++ b/Makefile.build
@@ -572,11 +572,11 @@ module Gramext = Gramlib__Gramext \
 module Grammar = Gramlib__Grammar" > $@
 
 gramlib__pack/gramlib__P%: gramlib/p% | gramlib__pack
-       cp -a $< $@
-       sed -e "1i # 1 \"$<\"" -i $@
+       printf "# 1 \"%s\"\n" $< > $@
+       cat $< >> $@
 gramlib__pack/gramlib__G%: gramlib/g% | gramlib__pack
-       cp -a $< $@
-       sed -e "1i # 1 \"$<\"" -i $@
+       printf "# 1 \"%s\"\n" $< > $@
+       cat $< >> $@

 
 # Specific rules for gramlib to pack it Dune / OCaml 4.08 style
 GRAMOBJS=$(addsuffix .cmo, $(GRAMFILES))
```
I did not follow precisely the whole thing. I hope I did not overlooked any action to do.